bootstrap/pkg/kfapp/coordinator/coordinator.go

```diff
@@ -241,121 +246,151 @@ func NewKfApp(options map[string]interface{}) (kftypes.KfApp, error) {
 		kfDef.Spec.Repo = path.Join(cacheDir, kftypes.KubeflowRepo)
-
-		pApp := GetKfApp(kfDef, nil)
-		return pApp, nil
+		kfDef.Spec.Project = options[string(kftypes.PROJECT)].(string)
+		kfDef.Spec.SkipInitProject = options[string(kftypes.SKIP_INIT_GCP_PROJECT)].(bool)
+		kfDef.Spec.UseBasicAuth = useBasicAuth
+		kfDef.Spec.UseIstio = useIstio
+		kfDef.Spec.PackageManager = packageManager
 	}
 
-	kfDef := &kfdefsv2.KfDef{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "KfDef",
-			APIVersion: "kfdef.apps.kubeflow.org/v1alpha1",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      appName,
-			Namespace: namespace,
-		},
-		Spec: kfdefsv2.KfDefSpec{
-			ComponentConfig: config.ComponentConfig{
-				Platform: platform,
-			},
-			Project:            project,
-			PackageManager:     packageManager,
-			UseBasicAuth:       useBasicAuth,
-			UseIstio:           useIstio,
-			EnableApplications: true,
-		},
-	}
-	configFileBuffer, configFileErr := getConfigFromCache(cacheDir, kfDef)
-	if configFileErr != nil {
-		log.Fatalf("could not get config file Error %v", configFileErr)
-	}
-	specErr := yaml.Unmarshal(configFileBuffer, kfDef)
-	if specErr != nil {
-		log.Errorf("couldn't unmarshal app.yaml. Error: %v", specErr)
-	}
+	// Disable usage report if requested
+	// TODO(jlewi): We should be able to get rid of this once we depend on this being
+	// configured in the config file.
 	disableUsageReport := options[string(kftypes.DISABLE_USAGE_REPORT)].(bool)
 	if disableUsageReport {
 		kfDef.Spec.Components = filterSpartakus(kfDef.Spec.Components)
 		delete(kfDef.Spec.ComponentParams, "spartakus")
 	}
 
-	kfDef.Name = appName
-	kfDef.Spec.AppDir = appDir
-	kfDef.Spec.Platform = platform
-	kfDef.Namespace = namespace
-	kfDef.Spec.Version = version
-	kfDef.Spec.Repo = path.Join(cacheDir, kftypes.KubeflowRepo)
-	kfDef.Spec.Project = options[string(kftypes.PROJECT)].(string)
-	kfDef.Spec.SkipInitProject = options[string(kftypes.SKIP_INIT_GCP_PROJECT)].(bool)
-	kfDef.Spec.UseBasicAuth = useBasicAuth
-	kfDef.Spec.UseIstio = useIstio
-	kfDef.Spec.PackageManager = packageManager
-	pApp := GetKfApp(kfDef, nil)
+	return NewKfAppFromKfDef(*kfDef)
+}
+
+// NewKfAppFromKfDef constructs the KfApp from the supplied KfDef.
+func NewKfAppFromKfDef(kfDef kfdefsv2.KfDef) (kftypes.KfApp, error) {
+	// Validate kfDef
+	errs := valid.NameIsDNSLabel(kfDef.Name, false)
+	if errs != nil && len(errs) > 0 {
+		return nil, &kfapis.KfError{
+			Code:    int(kfapis.INVALID_ARGUMENT),
+			Message: fmt.Sprintf(`invalid name due to %v`, strings.Join(errs, ", ")),
+		}
+	}
+
+	if _, err := os.Stat(kfDef.Spec.AppDir); os.IsNotExist(err) {
+		log.Infof("Creating directory %v", kfDef.Spec.AppDir)
+		appdirErr := os.MkdirAll(kfDef.Spec.AppDir, os.ModePerm)
+		if appdirErr != nil {
+			log.Errorf("couldn't create directory %v Error %v", kfDef.Spec.AppDir, appdirErr)
+			return nil, appdirErr
+		}
+	} else {
+		log.Infof("App directory exists %v", kfDef.Spec.AppDir)
```

Comment:
Should we exit here?
If user accidentally init targeting an existing deployment, config would be overwritten and not recoverable.

Reply:
I have refactored NewKfApp to split it into 3 well defined methods with clearer semantics

1. CreatekfDefFromOptions

   * This will parse command line options and construct a KfDef from the command line options
   * This will call the newly added backfileFromoptions method to backfill from KfDef from the many command line arguments we currently have (e.g. --email)
   * It will not attempt to write the KfDef to any file

1. CreateKfAppCfgFile 

    * This will attempt to write the KFDef to app.yaml
    * It will create the directory if needed
    * It will return an error if the app.yaml exists; this prevents the overwrite that Kunming pointed out

1. LoadKfAppCfgFile

     * Constructs a KfApp from an app.yaml file 
     * This is intended to be the main entry point once the config file exists.